feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/CallbackWrapper.kt

```diff
@@ -0,0 +1,281 @@
+        
+        fun fromCallback(cb: ()->Int): DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_f {
+            val callback: Runner_DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_f = object :  Runner_DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_f {
+                override fun invoke(ignored: Pointer?): Int {
+                    return cb();
+                }
+            }
+            val cb_wrap = DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_f_Native()
+            cb_wrap.run_callback = callback;
+            return DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_f(cb_wrap)
+        }
+    }
+}
+internal interface Runner_DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g: Callback {
+    fun invoke(ignored: Pointer?, arg0: Int ): Int
+}
+
+class DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g_Native: Structure(), Structure.ByValue {
+    @JvmField
+    internal var dataa: Pointer = Pointer(0L);
+    @JvmField
+    internal var run_callback: Runner_DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g
+        = object :  Runner_DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g {
+                override fun invoke(ignored: Pointer?, arg0: Int ): Int {
+                    throw Exception("ERROR NOT IMPLEMENTED")
+                }
+            }
+    @JvmField
+    internal var destructor: Pointer = Pointer(0L);
+
+    // Define the fields of the struct
+    override fun getFieldOrder(): List<String> {
+        return listOf("dataa", "run_callback", "destructor")
+    }
+}
+
+class DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g internal constructor (
+    internal val nativeStruct: DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g_Native) {
+    val dataa: Pointer = nativeStruct.dataa
+    val run_callback: Callback = nativeStruct.run_callback
+    val destructor: Pointer = nativeStruct.destructor
+
+    companion object {
+        val NATIVESIZE: Long = Native.getNativeSize(DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g_Native::class.java).toLong()
+        
+        fun fromCallback(cb: (Int)->Int): DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g {
+            val callback: Runner_DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g = object :  Runner_DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g {
+                override fun invoke(ignored: Pointer?, arg0: Int ): Int {
+                    return cb(arg0);
+                }
+            }
+            val cb_wrap = DiplomatCallback_CallbackWrapper_test_multiple_cb_args_diplomatCallback_g_Native()
+            cb_wrap.run_callback = callback;
```

Comment:
question: where is the data/dtor getting set here?

Reply:
In the previous line it's calling the internal constructor on the `Native` callback wrapper, which has the default value for `run_callback`, and a 0 pointer for `dataa`. The destructor is also set to a 0 pointer -- it won't do anything on the Rust side, since I think we decided Kotlin will do all the memory management for the callback structs. The direct assignment to `run_callback` here is overwriting the default value for this field.

Reply:
so wait, how does Rust tell Kotlin that it's no longer holding on to the callback?

I might be missing something in the design here.

Reply:
So I wasn't paying that much attention to the other PR but the callback is expanded into 
```rust
        #[no_mangle]
        extern "C" fn CallbackWrapper_test_multiple_cb_args(
            f: DiplomatCallback<i32>,
            g: DiplomatCallback<i32>,
        ) -> i32 {
            let f = move || unsafe {
                std::mem::transmute::<
                    unsafe extern "C" fn(*const c_void, ...) -> i32,
                    unsafe extern "C" fn(*const c_void) -> i32,
                >(f.run_callback)(f.data)
            };
            let g = move |arg0: i32| unsafe {
                std::mem::transmute::<
                    unsafe extern "C" fn(*const c_void, ...) -> i32,
                    unsafe extern "C" fn(*const c_void, i32) -> i32,
                >(g.run_callback)(g.data, arg0)
            };
            CallbackWrapper::test_multiple_cb_args(f, g)
        }
```
So the impl is only ever monomorphised into the closure `g`. As I understand it `g.data` and `g.run_callback` are copied into the closure and the closure is dropped after the call. But because they aren't owned pointers nothing is done to the backing data. And Kotlin will always wait until the method `CallbackWrapper.testMultipleCbArgs` is finished before it can GC the callback. 

I expect there would be a problem if a struct can have a callback as a field, or if we could wrap a callback in an opaque but this at least fails to compile because of no lifetime bounds on the `impl Fn`:
```
    #[diplomat::opaque]
    struct BadThing(Box<dyn Fn(i32) -> i32>);
    impl CallbackWrapper {
        ...
        pub fn test_opaque_callback(f: impl Fn(i32) -> i32) -> Box<BadThing> {
            Box::new(BadThing(Box::new(f)))
        }
    }
```

Reply:
But if we want callbacks as params or in opaques, then we need lifetime bounds afaics, and to add the callback struct to the edges of the the object.

Reply:
Good points -- indeed (as discussed offline too) we need to have Rust store a GC root to the callback wrapper object so that the JVM doesn't get rid of it until it's dropped on the Rust side too. We'll do this in a subsequent PR, but basically the mechanism will be to have the `data` field store a boxed JNI `GlobalRef` that references the callback wrapper object. Then in the `Drop`, Rust will drop the `GlobalRef`.